_blog/2024-02-08-uptime-monitoring-rust.mdx

```diff
@@ -0,0 +1,791 @@
+
+                             No data here :(</span>
+                     </div>
+                     {% else %}
+                     <div class="tooltip">
+                     🔴
+
+                             <span class="tooltiptext">{{timestamp.time}}
+
+                                     Uptime: {{timestamp.uptime_pct.unwrap()}}%</span>
+                     </div>
+
+                {% endmatch %}
+        {% endfor %}
+        </div>
+        <div>
+                <a href="/websites/{{log.alias}}" class="view-button">View</a>
+                <button hx-delete="/websites/{{log.alias}}"
+                        class="delete-button"
+                        hx-confirm="Are you sure you want to stop tracking this website?"
+                        hx-target="closest .menu"
+                        hx-swap="outerHTML">Delete</button>
+        </div>
+</div>
+{% endfor %}
+</div>
+{% endblock %}
+```
+
+As you can see here, we extend the `base.html` template and then loop through the websites we found earlier in our SQL query. We then display the timestamps as coloured circles depending on what the uptime percentage is (note that `None` means there’s a data gap).  Although we unwrap the uptime percentages here, we already match it beforehand to make sure it is a `Some` variant.
+
+You may have noticed we’re using tooltips to display the time of day that the request was taken as well as displaying the uptime percentage on the webpage. In the CSS file, we add styling so that when you hover over a circle, a tooltip will display the timestamp the circle represents as well as the exact uptime percentage.
+
+The single URL HTML webpage is also mostly the same, except we’re also adding an incident list:
+
+```html
+<!-- single_website.html -->
+{% extends "base.html" %}
+{% block content %}
+<h1>Shuttle Status Monitor</h1>
+<a href="/">Back to main page</a>
+<div class="website">
+        <h2 class="website-name">{{log.alias}} - {{log.url}}</h2>
+        <div>
+        Last 24  hours:
+        {% for timestamp in log.data %}
+                {% match timestamp.uptime_pct %}
+                     {% when Some with (100) %}
+                     <div class="tooltip">
+                             🟢
+                             <span class="tooltiptext">{{timestamp.time}}
+
+                                     Uptime: {{timestamp.uptime_pct.unwrap()}}%</span>
+                     </div>
+                     {% when None %}
+                     <div class="tooltip">
+                     ⚪
+                             <span class="tooltiptext">{{timestamp.time}}
+
+                             No data here :(</span>
+                     </div>
+                     {% else %}
+                     <div class="tooltip">
+                     🔴
+
+                             <span class="tooltiptext">{{timestamp.time}}
+
+                                     Uptime: {{timestamp.uptime_pct.unwrap()}}%</span>
+                     </div>
```

Comment:
Btw, do you use `prettier`?

Reply:
Not at the moment but this is probably worth a look. The HTML is not particularly well formatted

Reply:
It's a really nice little tool